enhancements/upgrade-bundle.md

```diff
@@ -0,0 +1,260 @@
+
+Let's consider a few scenarios: 
+
+  * Verifying operator upgrades
+    * To verify whether an operator upgrade to a desired version has been successful, the ClusterServiceVersion (CSV) CustomResource is a useful resource for providing such context to the end user. A given subscription has its status field and corresponding InstallPlan. If a ClusterServiceVersion (CSV) is created for that InstallPlan, then the CSV Status indicates that the operator has been successfully upgraded. We will ensure that the CSV has status `Installed` in the appropriate namespaces.  
+
+  * Handling subscription failures 
+    * If a subscription failed, there might not be an InstallPlan that's created for the new operator bundle. However, through the SubscriptionStatus, we can obtain the CurrentCSV and compare it with the desired operator version we wanted to upgrade to, and verify that they are not the same, and hence conclude there's been a failure. We then log a message that "run bundle-upgrade was not successful", and exit. 
+
+  * Operator fails to come up after creating subscription 
+    * If operator fails to come up after creating the subscription, it failed to create a second install plan and the SubscriptionStatus will indicate that. Therefore, we proceed to logging the error message and exiting. 
+
+  * What happens when a catalog source is not created?
+    * When a catalog source fails to create, it would show on the status of the CatalogSource. We log an error message, and exit. 
+
+  * Do we cleanup resources after failures or not? 
+    * We could offer flexibility to the operator developers to use a flag --skip-cleanup to indicate whether or not cleanup of resources is desired. This could enable the operator developers to capture all the logs and inspect the pod failures.
+
+  * Behavior of `run bundle-upgrade` in the presence of an OperatorGroup
+    * It is important to note that the presence of an OperatorGroup in the namespace can affect the how `run bundle-upgrade` operates. The presence of the OperatorGroup will take precedence. If the OperatorGroup does not support  version an error condition will occur. The `run bundle-upgrade` command will return with a non-zero exit status and log the errors. 
+
+  * Install Mode and OperatorGroups
+    * We support 3 modes:
+      * AllNamespaces
+      * OwnNamespace
+      * SingleNamespace=ns1
+        * NOTE: if ns1 is the same as the operator's namespace, a warning that "ns1 can not be the same as the operator's namespace" will be printed
+    * When `--install-mode` is not supplied, the selection algorithm is as follows:
+      * If an `OperatorGroup` already exists with the correct namespacing, use it.
+      * Else, create an `OperatorGroup` with the following preferences:
```

Comment:
There's a missing case here: The `OperatorGroup` exists and doesn't match the given `--install-mode`.

Reply:
In this case `bundle-upgrade` should error out.